### DISCUSSION TITLE
Decimate keyword Digital_input

### DISCUSSION THREAD

**Discussion by horsto** [pyControl/code#84] [Q&A] [answered]
I see that the `decimate` function in `Digital_input` has been removed in Version 1.8 (https://pycontrol.readthedocs.io/en/latest/about/release-notes/). The function seemed quite useful in my code since I am recording very long (over days) triggers from cameras that are sending TTLs at every exposure. Since those are perfectly regular, the decimate function saved some space and prevented crashes over the long pyControl recordings. Is there an alternative you recommend?

**Comment by ThomasAkam** [comment]
Hi @horsto,

Apologies for the inconvienience, I had not realised this functionallity was in active use as I was something I implemented very early in pycontrol development then never used subsequently in any of our experiments.  

We should certainly reimplement something for this if it is useful.  We might want to do it as a seperate class from the standard Digital_input though as it is a waste of pyboard processor resources to pass these events that occur at high rate and never interact with the state machine through the standard event processing pipeline.  I've created an issue (#86) for this and will implement something for the next release.

**Comment by horsto** [comment]
That makes sense - thanks for your reply! 
Do you have any other suggestions for recording "ultra long" (10-12 hours TTLs at 20 Hz) sessions? I am recording videos with Bonsai (bonsai-rx) and I am using pyControl to record sync pulses from my cameras while Bonsai is recording. I am mostly asking because pyControl does crash after about 4-6 hours of continuous recording (that does not surprise me given the amount of data that is written into the txt container). Any ideas?

**Comment by ThomasAkam** [comment, marked as answer]
I have implemented a `Frame_logger` class in the dev branch [here](https://github.com/pyControl/code/blob/dev/devices/frame_logger.py) which acts like a `Digital_input` but does not send the events to the state machine for processing, just saves their times to the log, and also has a decimate argument which works just like the old `Digital_input`.

It should work fine with the last release (v1.8.1).  To use it, just download the file `frame_logger.py` using the download button at the link above and put it in the `pyControl/devices` folder.  Then in your state machine instantiate a `Frame_logger` class as:

```python
logger = Frame_logger(pin='X1', rising_event='camera_frame', decimate=10)
```
Remember to include the event name in your `events` list. 

Let me know how you get on.

Also, please could you let me know what error message you get on crashes during long recordings (check the `error_log.txt` file if you don't see anything in the GUI log).

**Comment by horsto** [comment]
Thanks a lot, @ThomasAkam ! I will give it a try and report back on stability. I guess the `decimate` parameter state is retrievable after the recording has been done (to infer that indeed it has been subjected to decimate and to infer the correct data timestamps?)?

I do not see an `error_log.txt` - the program just crashes and has to be closed forcefully (Windows 10). 

Another possibility for me would be to call a python process once in a while to restart the process from Bonsa i, which I am using to record the videos (https://github.com/orgs/pyControl/discussions/85).

**Comment by ThomasAkam** [comment]
Currently the decimate parameter state is not recorded anywhere (as with the old Digital_input decimate functionallity).  If you want to record it you could do the following in your task file.

```python
v.decimate_param = 10
logger = Frame_logger(pin='X1', rising_event='camera_frame', decimate=v.decimate_param)

def run_start():
    print_variables('decimate_param')
```

**Reply by horsto** [reply]
Cool, thank you!
